mlir/lib/Gradient/Transforms/ConversionPatterns.cpp

```diff
@@ -178,6 +178,13 @@ struct BackpropOpPattern : public ConvertOpToLLVMPattern<BackpropOp> {
             // Register the previous functions as llvm globals (for Enzyme)
             insertEnzymeAllocationLike(rewriter, op->getParentOfType<ModuleOp>(), op.getLoc(),
                                        allocFn.getName(), freeFn.getName());
+
+            // Register free
+            // Allocation like fails to register the allocation function as the inverse of
+            // _mlir_memref_to_llvm_free.
```

Comment:
Small terminology nit: the term "inverse" is a bit confusing from an AD perspective. I'm used to "inverse" being associated with the differentiated version of a function, like how the derivative of `sin(x)` is `cos(x)`. From that perspective, you wouldn't differentiate a `malloc` call by inserting a `free`.

How Enzyme actually handles `free`s (roughly speaking) is by moving them outside of the primal (forward pass) and into the adjoint (reverse pass) because these buffers might be required in the adjoint. This behaviour doesn't really have an "inverse" relationship with how it handles `malloc`.

The overall result of this metadata is telling Enzyme:
1. `_mlir_memref_to_llvm_alloc`: when you see this, it allocates memory that you should then free using `_mlir_memref_to_llvm_free`
2. `_mlir_memref_to_llvm_free`: when you see this, it behaves *exactly* like `free`.

And both are required because Enzyme needs to know how to differentiate every function that could potentially impact the derivative.

It's a bit unfortunate that 2. isn't implied by 1. (but it makes sense because 1. does not necessarily imply 2.), but I think it would be good to have the comments express the purpose of these pieces of metadata as a whole.

Reply:
I think the `__enzyme_allocation_like` should handle the alloc, and `__enzyme_function_like` should handle the free. I would only add a third bit of metadata if necessary.